Std/Data/Array/Basic.lean

```diff
@@ -117,6 +117,75 @@ protected def maxI [ord : Ord α] [Inhabited α]
     (xs : Array α) (start := 0) (stop := xs.size) : α :=
   xs.minI (ord := ord.opposite) start stop
 
+
+/--
+Returns the first and all the rest of the elements of the array.
+-/
+protected def splitFirst (xs : Array α) (h : 0 < xs.size) : α × Array α :=
+  let head := xs.get ⟨0, h⟩
+  let tail := Subarray.mk xs 1 xs.size (Nat.succ_le_of_lt h) (Nat.le_refl _) |>.toArray
+  (head, tail)
+
+/--
+Returns the first and all the rest of the elements of the array, or `none` if it is empty.
+-/
+protected def splitFirst? (xs : Array α) : Option (α × Array α) :=
+  if h : 0 < xs.size then
+    let head := xs.get ⟨0, h⟩
+    let tail := Subarray.mk xs 1 xs.size (Nat.succ_le_of_lt h) (Nat.le_refl _) |>.toArray
+    some (head, tail)
+  else
+    none
+
+/--
+Returns the last and all the rest of the elements of the array, or `none` if it is empty.
+-/
+protected def splitLast? (xs : Array α) : Option (Array α × α) :=
+  if h1 : 0 < xs.size then
+    have h2 := Nat.sub_lt h1 (by decide)
+    let last := xs.get ⟨xs.size - 1, h2⟩
+    let heads := Subarray.mk xs 0 (xs.size - 1) (Nat.zero_le _) (Nat.le_of_lt h2) |>.toArray
+    some (heads, last)
+  else
+    none
+
+/--
+Returns the last and all the rest of the elements of the array.
+-/
+protected def splitLast (xs : Array α) (h : 0 < xs.size) : α × Array α :=
+  have h2 := Nat.sub_lt h (by decide)
+  let heads := xs.get ⟨xs.size - 1, h2⟩
+  let last := Subarray.mk xs 0 (xs.size - 1) (Nat.zero_le _) (Nat.le_of_lt h2) |>.toArray
+  (heads, last)
+
+/--
+Returns the first element of the array, or `none` if it is empty.
+-/
+protected def first? (xs : Array α) : Option α :=
+  if h : 0 < xs.size then
+    some (xs.get ⟨0, h⟩)
+  else
+    none
+
+/--
+Returns the first element of the array.
+-/
+protected def first (xs : Array α) (h : 0 < xs.size) : α :=
```

Comment:
This (and the others) should either be renamed to `front`, or we should add `last` as an alias for `back` or else it will be confusing.